File: lis.c
#include "mergesort.h"
#include "array.h"
#include <stdio.h>
#include <math.h>
#include <stdlib.h>

int compare(void *a, void *b)
{
	return ((Item *) a)->val - ((Item *) b)->val;
}

Item *bin_search(Array *X, int start, int end, int value)
{
	if (end - start <= 1)
	{
		return X->items[start].val < value ? &X->items[start] : NULL;
	}
	int half = (end - start) / 2;
	Item *l = bin_search(X, start, start + half, value);
	Item *r = bin_search(X, start + half + 1, end, value);
	if (!l)
	{
		return r;
	}
	else if (!r)
	{
		return l;
	}
	else
	{
		return r;
	}
}

void combine_lis(Array *X, int i, int j, int m_e, int *pred)
{
	int max_ind = 0, max_len = 0, longest_LIS_len = 0;
	int n = j - i + 1;
	Array *Knuth = Array_new();
	Array_append(Knuth, 0, 0);
	int *is_present = calloc(X->len, sizeof(int));
	for (size_t r = 0; r < n; r++)
	{
		int k = pred[X->items[r].pos];
		int j = is_present[k];
		if (X->items[r].val <= m_e)
		{
			if (longest_LIS_len <= j + 1)
			{
				longest_LIS_len = j + 1;
			}
		}
		else
		{
			if (k == 0)
			{
				j = longest_LIS_len;
			}
			else
			{
				if (j == 0)
				{
					Item *p = bin_search(X, longest_LIS_len, max_len, X->items[r].val);
					j = p->pos;
				}
			}
		}
		is_present[Knuth->items[j + 1].pos] = 0;
		pred[X->items[r].pos] = Knuth->items[j].pos;
		Knuth->items[j + 1].val = X->items[r].val;
		Knuth->items[j + 1].pos = X->items[r].pos;
		is_present[X->items[r].pos] = j + 1;
		if (max_len <= j + 1)
		{
			max_ind = X->items[r].pos;
			max_len = j + 1;
		}
	}
	Array_free(Knuth);
	free(is_present);
}

void divide_lis(Array *X, int i, int j, int *pred)
{
	if (i == j)
	{
		pred[i] = 0;
	}
	else if (j > i)
	{
		int n = j - i + 1;
		int x = (int) floor((j - i + 1) / 2.0);
		Array *Xsorted = Array_copy(X);
		// qsort(Xsorted->items, X->len - 1, sizeof(Item), (__compar_fn_t) &compare);
		int m_e = Xsorted->items[i + x].val;
		Array *Y1 = Array_new();
		Array *Y2 = Array_new();
		for (size_t k = 0; k < n; k++)
		{
			if (X->items[k].val < m_e)
			{
				Array_append(Y1, X->items[k].val, X->items[k].pos);
			}
			else
			{
				Array_append(Y2, X->items[k].val, X->items[k].pos);
			}
		}
		divide_lis(Y1, i, i + x - 1, pred);
		divide_lis(Y2, x + i, j, pred);
		combine_lis(X, i, j, m_e, pred);
		Array_free(Y1);
		Array_free(Y2);
	}
}

int main(void)
{
	Array *X = Array_of(10, 8, 9, 5, 2, 3, 7, 10, 4, 1, 6);
	int *pred = calloc(X->len, sizeof(int));
	divide_lis(X, 0, X->len - 1, pred);
}

File: lis.py
import math

def merge(arr, l, m, r):
	n1 = m - l + 1
	n2 = r - m

	# create temp arrays
	L = [0] * (n1)
	R = [0] * (n2)

	# Copy data to temp arrays L[] and R[]
	for i in range(0, n1):
		L[i] = arr[l + i]

	for j in range(0, n2):
		R[j] = arr[m + 1 + j]

	# Merge the temp arrays back into arr[l..r]
	i = 0     # Initial index of first subarray
	j = 0     # Initial index of second subarray
	k = l     # Initial index of merged subarray

	while i < n1 and j < n2:
		if L[i] <= R[j]:
			arr[k] = L[i]
			i += 1
		else:
			arr[k] = R[j]
			j += 1
		k += 1

	# Copy the remaining elements of L[], if there
	# are any
	while i < n1:
		arr[k] = L[i]
		i += 1
		k += 1

	# Copy the remaining elements of R[], if there
	# are any
	while j < n2:
		arr[k] = R[j]
		j += 1
		k += 1

# l is for left index and r is right index of the
# sub-array of arr to be sorted

def mergeSort(arr, l, r):
	if l < r:

		# Same as (l+r)//2, but avoids overflow for
		# large l and h
		m = l+(r-l)//2

		# Sort first and second halves
		mergeSort(arr, l, m)
		mergeSort(arr, m+1, r)
		merge(arr, l, m, r)

def divide_lis(X: list, i, j, pred):
	if i == j:
		pred[i] = 0
	elif j > i:
		n = j - i + 1
		x = math.floor((j - i + 1) / 2.0)
		Xsorted = X.copy()
		mergeSort(Xsorted, 0, len(Xsorted) - 1)
		m_e = Xsorted[i + x].val
		Y1 = []
		Y2 = []
		for k in range(0,n):
			if X[k] < m_e:
				Y1.append(X[k])
			else:
				Y2.append(X[k])
		divide_lis(Y1, i, i + x - 1, pred)
		divide_lis(Y2, x + i, j, pred)
		combine_lis(X, i, j, m_e, pred)

def combine_lis(X, i, j, m_e, pred):
	max_ind = 0, max_len = 0, longest_LIS_len = 0
	n = j - i + 1
	Knuth = []
	Knuth.append( 0, 0)
	int *is_present = []
	for r in range(0, n):
		k = pred[X[r].pos]
		j = is_present[k]
		if X[r].val <= m_e:
			if (longest_LIS_len <= j + 1):
				longest_LIS_len = j + 1
		else:
			if k == 0:
				j = longest_LIS_len
			else:
				if j == 0:
					Item *p = bin_search(X, longest_LIS_len, max_len, X->items[r].val)
					j = p->pos
		is_present[Knuthitems[j + 1].pos] = 0
		pred[X[r].pos] = Knuth[j].pos
		Knuth[j + 1].val = X[r].val
		Knuth[j + 1].pos = X[r].pos
		is_present[Xitems[r].pos] = j + 1
		if max_len <= j + 1:
			max_ind = X[r].pos
			max_len = j + 1
